From 79b52a62c24680ff8d22f8b999519a88f216b184 Mon Sep 17 00:00:00 2001
From: Truls Asheim <truls@asheim.dk>
Date: Mon, 9 Jun 2025 15:49:45 +0200
Subject: [PATCH] Free OpenSSL engine only after signing is complete

This patch adds support for signing using engines where the handles
the actual signing. This is done by keeping the engine allocated
throughout the signing operation, both when acquiring the private key
and when signing the EFI variable. Previously, the engine would be
freed after using it to acquire the private key causing subsequent
signing operations that depend on the engine to fail.

Signed-off-by: Truls Asheim <truls@asheim.dk>
---
 include/openssl_sign.h |  2 +-
 lib/openssl_sign.c     | 56 ++++++++++++++++++++++--------------------
 2 files changed, 31 insertions(+), 27 deletions(-)

diff --git a/include/openssl_sign.h b/include/openssl_sign.h
index 136ad75..5ca6ffb 100644
--- a/include/openssl_sign.h
+++ b/include/openssl_sign.h
@@ -7,4 +7,4 @@ int
 sign_efi_var_ssl(char *payload, int payload_size, EVP_PKEY *pkey, X509 *cert,
 		 unsigned char **sig, int *sigsize);
 EVP_PKEY *
-read_private_key(char *engine, char *keyfile);
+read_private_key(ENGINE *engine, char *keyfile);
diff --git a/lib/openssl_sign.c b/lib/openssl_sign.c
index 714ce1a..65dcdb9 100644
--- a/lib/openssl_sign.c
+++ b/lib/openssl_sign.c
@@ -36,6 +36,7 @@ sign_efi_var(char *payload, int payload_size, char *keyfile, char *certfile,
 	     unsigned char **sig, int *sigsize, char *engine)
 {
 	int ret;
+	ENGINE *e = NULL;
 
 	ERR_load_crypto_strings();
 	OpenSSL_add_all_digests();
@@ -59,16 +60,39 @@ sign_efi_var(char *payload, int payload_size, char *keyfile, char *certfile,
 		fprintf(stderr, "error reading certificate %s\n", certfile);
 		return 1;
 	}
+	if (engine) {
+		ENGINE_load_builtin_engines();
+		e = ENGINE_by_id(engine);
+
+		if (!e) {
+			fprintf(stderr, "Failed to load engine: %s\n", engine);
+			ERR_print_errors_fp(stderr);
+			ret = 1;
+			goto cert_free;
+		}
+		if (!ENGINE_init(e)) {
+			fprintf(stderr, "Failed to initialize engine %s\n", engine);
+			ERR_print_errors_fp(stderr);
+			ret = 1;
+			goto engine_free;
+		}
+	}
 
-	EVP_PKEY *pkey = read_private_key(engine, keyfile);
+	EVP_PKEY *pkey = read_private_key(e, keyfile);
 	if (!pkey) {
 		ERR_print_errors_fp(stdout);
 		fprintf(stderr, "error reading private key %s\n", keyfile);
-		return 1;
+		ret = 1;
+		goto engine_free;
 	}
 	ret = sign_efi_var_ssl(payload, payload_size, pkey, cert,
 			       sig, sigsize);
 	EVP_PKEY_free(pkey);
+
+engine_free:
+	ENGINE_finish(e);
+	ENGINE_free(e);
+cert_free:
 	X509_free(cert);
 
 	return ret;
@@ -109,45 +133,25 @@ static int ui_read(UI *ui, UI_STRING *uis)
 }
 
 static EVP_PKEY *
-read_engine_private_key(char *engine, char *keyfile)
+read_engine_private_key(ENGINE *engine, char *keyfile)
 {
 	UI_METHOD *ui;
-	ENGINE *e;
 	EVP_PKEY *pkey = NULL;
 
-	ENGINE_load_builtin_engines();
-	e = ENGINE_by_id(engine);
-
-	if (!e) {
-		fprintf(stderr, "Failed to load engine: %s\n", engine);
-		ERR_print_errors_fp(stderr);
-		return NULL;
-	}
-
 	ui = UI_create_method("sbsigntools");
 	if (!ui) {
 		fprintf(stderr, "Failed to create UI method\n");
 		ERR_print_errors_fp(stderr);
-		goto out_free;
+		return NULL;
 	}
 	UI_method_set_reader(ui, ui_read);
 
-	if (!ENGINE_init(e)) {
-		fprintf(stderr, "Failed to initialize engine %s\n", engine);
-		ERR_print_errors_fp(stderr);
-		goto out_free;
-	}
-
-	pkey = ENGINE_load_private_key(e, keyfile, ui, NULL);
-	ENGINE_finish(e);
-
- out_free:
-	ENGINE_free(e);
+	pkey = ENGINE_load_private_key(engine, keyfile, ui, NULL);
 	return pkey;
 }
 
 EVP_PKEY *
-read_private_key(char *engine, char *keyfile)
+read_private_key(ENGINE *engine, char *keyfile)
 {
 	if (engine)
 		return read_engine_private_key(engine, keyfile);
-- 
2.43.0

